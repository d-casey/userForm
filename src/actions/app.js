import {
  RESET_FORM_STAGE,
  UPDATE_STAGE
} from '../constants/actions'

export const updateStage = stage => ({
  type: UPDATE_STAGE,
  payload: {
    stage
  }
})

export const resetFormStage = () => ({
  type: RESET_FORM_STAGE
})
