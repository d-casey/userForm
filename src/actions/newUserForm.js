import {
  RESET_USER_FORM,
  SUBMIT_NEW_USER_FORM
} from '../constants/actions'

export const submitNewUserForm = formFields => ({
  type: SUBMIT_NEW_USER_FORM,
  payload: formFields
})

export const resetUserForm = () => ({
  type: RESET_USER_FORM
})
