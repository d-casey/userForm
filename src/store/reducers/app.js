import {
  UPDATE_STAGE,
  RESET_FORM_STAGE,
  SET_FORM_COMPLETION,
  RESET_APP
} from '../constants/actions'

const initialState = {
  stage: 0,
  formCompleted: false
}

export default (state = initialState, action) => {
  switch (action.type) {
    case UPDATE_STAGE:
      return {
        ...state,
        stage: action.payload.stage,
      }

    case RESET_FORM_STAGE:
      return {
        ...state,
        stage: 0
      }

    case SET_FORM_COMPLETION:
      return {
        ...state,
        formCompleted: true
      }

    case RESET_APP:
      return {
        ...state,
        formCompleted: false,
        stage: 0
      }

    default:
      return state
  }
}
