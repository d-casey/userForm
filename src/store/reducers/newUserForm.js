import {
  UPDATE_USER_FIELDS,
  RESET_USER_FORM
} from '../constants/actions'

export const initialState = {
  name: "",
  role: "",
  email: "",
  password: "",
  isSubmitted: false
}

export default (state = initialState, action) => {
  switch (action.type) {
    case UPDATE_USER_FIELDS:
      return {
        ...state,
        name: action.payload.name,
        role: action.payload.role,
        email: action.payload.email,
        password: action.payload.password,
        isSubmitted: true
      }

    case RESET_USER_FORM:
      return initialState

    default:
      return state
  }
}
