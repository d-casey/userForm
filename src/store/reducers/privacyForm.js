import {
  RESET_PRIVACY_FORM,
  SUBMIT_PRIVACY_FORM
} from '../constants/actions'

const initialState = {
  receive_updates: true,
  receive_communication: false,
  isSubmitted: false
}

export default (state = initialState, action) => {
  switch (action.type) {
    case SUBMIT_PRIVACY_FORM:
      return {
        ...state,
        receive_updates: action.payload.receive_updates,
        receive_communication: action.payload.receive_communication,
        isSubmitted: true
      }

      case RESET_PRIVACY_FORM:
        return initialState

    default:
      return state
  }
}
