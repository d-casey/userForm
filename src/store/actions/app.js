import {
  RESET_FORM_STAGE,
  UPDATE_STAGE,
  LOGGING_DETAILS,
  SET_FORM_COMPLETION,
  RESET_APP
} from '../constants/actions'

import { resetUserForm, resetPrivacyForm } from '../../store/actions'

export const updateStage = stage => ({
  type: UPDATE_STAGE,
  payload: {
    stage
  }
})

export const resetFormStage = () => ({
  type: RESET_FORM_STAGE
})

export const resetAllForms = () => dispatch => {
  dispatch(resetUserForm())
  dispatch(resetPrivacyForm())
}

export const setFormCompletion = () => ({
  type: SET_FORM_COMPLETION
})

export const resetApp = () => ({
  type: RESET_APP
})

export const loggingDetails = () => ({
  type: LOGGING_DETAILS
})

export const logSubmittedFormDetails = (formDetails) => dispatch => {
  dispatch(loggingDetails())
  console.log('formDetails: ', formDetails)
  dispatch(resetAllForms())
  dispatch(setFormCompletion())
}
