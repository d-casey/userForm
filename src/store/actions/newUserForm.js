import {
  RESET_USER_FORM,
  UPDATE_USER_FIELDS
} from '../constants/actions'

import { updateStage } from '../actions'

export const updateUserFields = formFields => ({
  type: UPDATE_USER_FIELDS,
  payload: formFields
})

export const submitNewUserForm = formFields => dispatch => {
  dispatch(updateUserFields(formFields))
  dispatch(updateStage(1))
}

export const resetUserForm = () => ({
  type: RESET_USER_FORM
})
