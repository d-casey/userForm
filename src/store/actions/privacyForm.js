import {
  SUBMIT_PRIVACY_FORM,
  RESET_PRIVACY_FORM
} from '../constants/actions'

import { updateStage } from '../actions'

export const updatePrivacyFields = formFields => ({
  type: SUBMIT_PRIVACY_FORM,
  payload: formFields
})

export const submitPrivacyForm = formFields => dispatch => {
  dispatch(updatePrivacyFields(formFields))
  dispatch(updateStage(2))
}

export const resetPrivacyForm = () => ({
  type: RESET_PRIVACY_FORM
})
