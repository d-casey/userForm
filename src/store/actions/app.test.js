import * as appActions from './app'
import * as newUserFormActions from './newUserForm'
import * as privacyFormActions from './privacyForm'

describe('app actions', () => {
  describe('updateStage', () => {
    it('sends the stage as the payload', () => {
      expect(appActions.updateStage({ stage: 1 })).toMatchSnapshot()
    })
  })

  describe('resetFormStage', () => {
    it('sends the actions', () => {
      expect(appActions.resetFormStage()).toMatchSnapshot()
    })
  })

  describe('setFormCompletion', () => {
    it('sets the form completion', () => {
      expect(appActions.setFormCompletion()).toMatchSnapshot()
    })
  })

  describe('resetApp', () => {
    it('resets the app', () => {
      expect(appActions.resetApp()).toMatchSnapshot()
    })
  })

  describe('loggingDetails', () => {
    it('says that we are logging the details', () => {
      expect(appActions.loggingDetails()).toMatchSnapshot()
    })
  })

  describe('resetAllForms', () => {
    let dispatch
    beforeEach(() => {
      dispatch = jest.fn()
      newUserFormActions.resetUserForm = jest.fn(() => 'RESET_USER_FORM')
      privacyFormActions.resetPrivacyForm = jest.fn(() => 'RESET_PRIVACY_FORM')
      appActions.resetAllForms()(dispatch)
    })

    test('it calls dispatch two times', () => {
      expect(dispatch).toHaveBeenCalledTimes(2)
    })

    it('calls dispatch with the resetUserForm actions', () => {
      expect(dispatch).toBeCalledWith('RESET_USER_FORM')
    })

    it('calls dispatch with the resetPrivacyForm actions', () => {
      expect(dispatch).toBeCalledWith('RESET_PRIVACY_FORM')
    })
  })

  describe('logSubmittedFormDetails', () => {
    let dispatch
    let formDetails = 'FORM_DETAILS'
    beforeEach(() => {
      dispatch = jest.fn()
      jest.spyOn(global.console, 'log')
      appActions.logSubmittedFormDetails(formDetails)(dispatch)
    })

    test('it calls dispatch three times', () => {
      expect(dispatch).toHaveBeenCalledTimes(3)
    })

    it('calls dispatch with the loggingDetails actions', () => {
      expect(dispatch).toBeCalledWith({ 'type': 'LOGGING_DETAILS' })
    })

    it('calls console.log with the form details', () => {
      expect(console.log).toBeCalledWith('formDetails: ', 'FORM_DETAILS')
    })

    it('calls dispatch with the setFormCompletion actions', () => {
      expect(dispatch).toBeCalledWith({ 'type': 'SET_FORM_COMPLETION' })
    })
  })
})
