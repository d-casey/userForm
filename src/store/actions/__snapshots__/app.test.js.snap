// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app actions loggingDetails says that we are logging the details 1`] = `
Object {
  "type": "LOGGING_DETAILS",
}
`;

exports[`app actions resetApp resets the app 1`] = `
Object {
  "type": "RESET_APP",
}
`;

exports[`app actions resetFormStage sends the actions 1`] = `
Object {
  "type": "RESET_FORM_STAGE",
}
`;

exports[`app actions setFormCompletion sets the form completion 1`] = `
Object {
  "type": "SET_FORM_COMPLETION",
}
`;

exports[`app actions updateStage sends the stage as the payload 1`] = `
Object {
  "payload": Object {
    "stage": Object {
      "stage": 1,
    },
  },
  "type": "UPDATE_STAGE",
}
`;
